gilbreth_perception/src/segmentation_node.cpp

```diff
@@ -55,45 +58,50 @@ void cloudCb(const sensor_msgs::PointCloud2ConstPtr &cloud_msg) {
-  sor.setInputCloud(scene_filtered);
-  sor.setLeafSize(down_sample, down_sample, down_sample);
-  sor.filter(*scene);
-
-  // Cluster Extraction
-  pcl::search::KdTree<pcl::PointXYZ>::Ptr tree(new pcl::search::KdTree<pcl::PointXYZ>());
-  tree->setInputCloud(scene);
-  std::vector<pcl::PointIndices> cluster_indices;
-  pcl::EuclideanClusterExtraction<pcl::PointXYZ> ec;
-  ec.setClusterTolerance(0.1); // 10cm
-  ec.setMinClusterSize(80);
-  ec.setMaxClusterSize(25000);
-  ec.setSearchMethod(tree);
-  ec.setInputCloud(scene);
-  ec.extract(cluster_indices);
-
-  std::cerr << "Cluster number is : " << cluster_indices.size() << std::endl;
-
-  // Publish cluster clouds
-  pcl::PointIndices::Ptr indices(new pcl::PointIndices);
-  pcl::ExtractIndices<pcl::PointXYZ> extract;
-  pcl::PointCloud<pcl::PointXYZ>::Ptr cluster_cloud(new pcl::PointCloud<pcl::PointXYZ>());
-  sensor_msgs::PointCloud2 output;
-
-  for (int i = 0; i < cluster_indices.size(); i++) {
-    *indices = cluster_indices[i];
-    extract.setInputCloud(scene);
-    extract.setIndices(indices);
-    extract.setNegative(false);
-    extract.filter(*cluster_cloud);
-
-    pcl::toROSMsg(*cluster_cloud, output);
-    output.header.stamp = cloud_msg->header.stamp;
-    pub.publish(output);
-
-    std::cerr << "Publish cluster " << i << std::endl;
+  if (scene_filtered->points.size() > 0) {
+    // Downsample scene
+    pcl::VoxelGrid<pcl::PointXYZ> sor;
+    sor.setInputCloud(scene_filtered);
+    sor.setLeafSize(down_sample, down_sample, down_sample);
+    sor.filter(*scene);
+
+    // Cluster Extraction
+    pcl::search::KdTree<pcl::PointXYZ>::Ptr tree(new pcl::search::KdTree<pcl::PointXYZ>());
+    tree->setInputCloud(scene);
+    std::vector<pcl::PointIndices> cluster_indices;
+    pcl::EuclideanClusterExtraction<pcl::PointXYZ> ec;
+    ec.setClusterTolerance(0.1); // 10cm
+    ec.setMinClusterSize(80);
+    ec.setMaxClusterSize(25000);
+    ec.setSearchMethod(tree);
+    ec.setInputCloud(scene);
+    ec.extract(cluster_indices);
+    if (print_detailed_info) {
+      std::cerr << "Cluster number is : " << cluster_indices.size() << std::endl;
+    }
+
+    // Publish cluster clouds
+    if (cluster_indices.size() > 0) {
+      pcl::PointIndices::Ptr indices(new pcl::PointIndices);
+      pcl::ExtractIndices<pcl::PointXYZ> extract;
+      pcl::PointCloud<pcl::PointXYZ>::Ptr cluster_cloud(new pcl::PointCloud<pcl::PointXYZ>());
+      sensor_msgs::PointCloud2 output;
+
+      for (int i = 0; i < cluster_indices.size(); i++) {
+        *indices = cluster_indices[i];
+        extract.setInputCloud(scene);
+        extract.setIndices(indices);
+        extract.setNegative(false);
+        extract.filter(*cluster_cloud);
+
+        pcl::toROSMsg(*cluster_cloud, output);
+        output.header.stamp = cloud_msg->header.stamp;
+        pub.publish(output);
+        if (print_detailed_info) {
+          std::cerr << "Publish cluster " << i + 1 << std::endl;
+        }
+      }
+    }
```

Comment:
In general, I try to avoid too many levels of nested if statements,  whenever possible I prefer to do as follows:
```
if (condition1_valid == true)
{
  // ... do something meaningful
}
else
{
  // ... error handling procedure if applicable
  return;
}

if (condition2_valid == true)
{
  // ... do something meaningful
}
else
{
  // ... error handling procedure if applicable
  return;
}
```
In my opinion it helps with code readability, however no change is needed on this one.